snap-jp2-reader/src/main/java/org/esa/snap/jp2/reader/metadata/Jp2XmlMetadata.java

```diff
@@ -152,38 +155,46 @@ public Point2D getOrigin() {
+     *  Keep in mind that there are JP2 files with values like ["  -0.000004629629630 0", "0    0.000004629629630"] for offsetVector
+     * @return
+     */
     public double getStepX() {
         String[] values = getAttributeValues(JP2ProductReaderConstants.TAG_OFFSET_VECTOR);
+        String[] result;
         if (values != null) {
-            // SNAP-1400 (https://senbox.atlassian.net/browse/SNAP-1400)
-            // return Double.parseDouble(values[0].split(" ")[0]);
-            // there are JP2 files with values like ["  -0.000004629629630 0", "0    0.000004629629630"]
-            if (!getCrsGeocoding().equals("EPSG::4326")) {
-                final String[] result = Arrays.stream(values[0].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
-                return Double.parseDouble(result[0]);
+            if (!isReversedAxisOrder()) {
+                result = Arrays.stream(values[0].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
             }
             else {
-                final String[] result = Arrays.stream(values[1].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
-                return Double.parseDouble(result[1]);
+                result = Arrays.stream(values[1].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
             }
+            return Double.parseDouble(result[1]) != 0 ? Double.parseDouble(result[1]) : Double.parseDouble(result[0]);
         }
         return 0;
     }
 
+    /**
+     * SNAP-1400 (https://senbox.atlassian.net/browse/SNAP-1400)
+     * SNAP-3459 (https://senbox.atlassian.net/browse/SNAP-3459) - Do not rely on EPSG and check the actual axis name order;
+     *                                                             also the values within offsetVector should be considered depending on their logic.
+     *  Keep in mind that there are JP2 files with values like ["  -0.000004629629630 0", "0    0.000004629629630"] for offsetVector
+     * @return
+     */
     public double getStepY() {
         String[] values = getAttributeValues(JP2ProductReaderConstants.TAG_OFFSET_VECTOR);
+        String[] result;
         if (values != null) {
-            // SNAP-1400 (https://senbox.atlassian.net/browse/SNAP-1400)
-            //return Double.parseDouble(values[1].split(" ")[1]);
-            // there are JP2 files with values like ["  -0.000004629629630 0", "0    0.000004629629630"]
-            if (!getCrsGeocoding().equals("EPSG::4326")) {
-                final String[] result = Arrays.stream(values[1].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
-                return Double.parseDouble(result[1]);
+            if (!isReversedAxisOrder()) {
```

Comment:
This if-else can be replaced by:
```
            int index = isReversedAxisOrder() ? 0 : 1;
            result = Arrays.stream(values[index].split(" ")).filter(p -> !StringUtils.isEmpty(p)).map(String::trim).toArray(String[]::new);
```
Looks cleaner and is easier to read in my opinion. The other locations can be enahnced similarly.

Reply:
Code refactoring done.